### ISSUE TITLE
Cannot cast object 'org.scoverage.ScoverageExtension_Decorated@70bf1425'

### ISSUE THREAD

**Issue by eyalroth** [scoverage/gradle-scoverage#87]
In a multi-module project, defining the aggregation task must be made in the `build.gradle` and cannot be extracted to a separate script file.

The following configuration:

`build.gradle`:
```groovy
plugins {
    id "org.scoverage" version "2.5.0" apply false
}
apply from: new File(rootProject.projectDir, "gradle/scoverage.gradle")
```

`gradle/scoverage.gradle`:
```groovy
buildscript {
    repositories {
        maven {
            url "https://plugins.gradle.org/m2/"
        }
    }

    dependencies {
        // necessary for resolving 'org.scoverage.ScoverageAggregate'
        classpath group: 'gradle.plugin.org.scoverage', name: 'gradle-scoverage', version: '2.5.0'
    }
}

task aggregateScoverage(type: org.scoverage.ScoverageAggregate)
```

Will cause this exception upon execution of `aggregateScoverage`:
```
 Caused by: org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object 'org.scoverage.ScoverageExtension_Decorated@70bf1425' with class 'org.scoverage.ScoverageExtension_Decorated' to class 'org.scoverage.ScoverageExtension'
        at org.scoverage.ScoveragePlugin.extensionIn(ScoveragePlugin.groovy:22)
        at org.scoverage.ScoverageAggregate.exec(ScoverageAggregate.groovy:12)
```

There are two problems here:
1. The incorrect reading of properties from within tasks. `ScoveragePlugin#extensionIn` is not the correct way of reading configuration from the extension. See Gradle's official guide on [how to map extension properties to task properties](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:mapping_extension_properties_to_task_properties).
2. Users shouldn't be bothered with defining the aggregation task on their own, which requires the entire `buildscript` section unless the definition is in the main `build.gradle` file. The task should be defined by default upon applying the plugin, much like the other tasks.

**Comment by maiflai** [comment]
Thanks - I'll take a look at the best practice link.

On the second point - this was originally to avoid to creating an aggregation task for every project. My personal use case was to create one overall report, so defining the task at the root project seemed straightforward.

Is this related to your java+scala work? I suppose that in a mixed project it might be useful to aggregrate at different levels?

**Comment by maiflai** [comment]
Yes - but that then means that the trivial single project case now runs an unnecessary aggregation?

I'll see what I can come up with.

**Comment by eyalroth** [comment]
@maiflai It won't run it if the user won't ask for it.

**Comment by eyalroth** [comment]
@maiflai I'm actually working on fixing this at the moment, alongside removing the requirement of explicitly declaring the scoverage scalac dependencies.

**Comment by maiflai** [comment]
ah - specifying the scoverage dependencies is a deliberate choice. I don't want to release a new version of the gradle plugin for every upstream release.

I also don't want to automatically take upstream releases.

**Comment by eyalroth** [comment]
@maiflai That's totally fine. The gradle plugin should use `1.3.1` by default, but allow the version to be configurable via an extension property; for instance:
```groovy
scoverage {
  scoverageVersion = '1.3.9'
}
```

**Comment by maiflai** [comment]
Has this been resolved by the 3.0.0 release please?

Thanks,
Stu

**Comment by eyalroth** [comment]
> Has this been resolved by the 3.0.0 release please?
> 
> Thanks,
> Stu

Yes :)
